mozregression/main.py

```diff
@@ -308,6 +308,101 @@ def check_mozregression_version():
+    AppArmor.  This might result in crashes on older builds, and in degraded
+    sandbox behavior.  It is fixed with an AppArmor profile that allows the
+    syscall to proceed, but this is path dependant on the binary we download
+    and needs to be installed at a system level, so we can only advise people
+    of the situation.
+
+    The following sys entry should be enough to verify whether it is blocked or
+    not, but the Ubuntu security team recommend cross-checking with actual
+    syscall.  This code is a simplification of how Firerox does it, cf
+    https://searchfox.org/mozilla-central/rev/23efe2c8c5b3a3182d449211ff9036fb34fe0219/security/sandbox/linux/SandboxInfo.cpp#114-175
+    and has been the most reliable way so far (shell with unshare would not
+    reproduce EPERM like we want).
+    """
+
+    apparmor_file = "/proc/sys/kernel/apparmor_restrict_unprivileged_userns"
+    if not os.path.isfile(apparmor_file):
+        return
+
+    with open(apparmor_file, "r") as f:
+        if f.read().strip() != "1":
+            return
+
+    import ctypes
+    import errno
+    import platform
+    import signal
+
+    # Values are from
+    # https://github.com/hrw/syscalls-table/tree/163e238e4d7761fcf6ac500aad92d53ac88d663a/system_calls/tables
+    # imported from linux kernel headers
+    SYS_clone = {
+        "i386": 120,
+        "x32": 1073741880,
+        "x86_64": 56,
+        "arm": 120,
+        "armv7l": 120,
+        "arm64": 220,
+        "aarch64": 220,
+        "aarch64_be": 220,
+        "armv8b": 220,
+        "armv8l": 220,
+    }.get(platform.machine())
+    if not SYS_clone:
+        LOG.warning(
+            "Unprivileged user namespaces might be disabled, but unsupported platform? {}".format(
+                platform.machine()
+            )
+        )
+        return
+
+    libc = ctypes.CDLL(None, use_errno=True)
+
+    LOG.warning(
+        "Unprivileged user namespaces might be disabled. Checking clone() + unshare() syscalls ..."
+    )
+
+    pid = libc.syscall(SYS_clone, signal.SIGCHLD.value | os.CLONE_NEWUSER, None, None, None, None)
+    print("pid", pid)
+
+    if pid == 0:
+        # Child side ...
+        rv = libc.unshare(os.CLONE_NEWPID)
+        _errno = ctypes.get_errno()
+        print("rv=", rv)
+        if rv < 0:
+            print("rv", rv, "errno", _errno)
+            sys.exit(_errno)
+        sys.exit(0)
+    else:
+        print("waiting", pid)
+        (pid, statuscode) = os.waitpid(pid, 0)
+        exitcode = os.waitstatus_to_exitcode(statuscode)
+        print("waited", pid, "finshed with", statuscode, " => ", exitcode)
+
+        if exitcode == 0:
+            return
+
+        if exitcode == errno.EPERM:
+            LOG.warning(
+                "Unprivileged user namespaces are disabled. This is likely because AppArmor policy "
+                "change. Please refer to XXX to learn how to setup AppArmor so that MozRegression "
```

Comment:
@zzzeid This is left `XXX` on purpose: where and how do we want to inform user? There's a pending SUMO page but it will be tailored to a different usecase, we might want something else for MozRegression. There's no Wiki setup.